# Exploring Hacker News Posts

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

You can find the data set [here](https://www.kaggle.com/hacker-news/hacker-news-posts), but note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions. Below are descriptions of the columns:

* id: The unique identifier from Hacker News for the post
* title: The title of the post
* url: The URL that the posts links to, if it the post has a URL
* num_points: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
* num_comments: The number of comments that were made on the post
* author: The username of the person who submitted the post
* created_at: The date and time at which the post was submitted

In [1]:
from csv import reader
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
data_as_list = list(read_file)
hn_header = data_as_list[0]
hn = data_as_list[1:]
print(hn_header)
print(hn[:6])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12'], ['10482257', 'Ti

We're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question and Show HN posts to show the Hacker News community a project, product, or just generally something interesting. 

In [2]:
#Create three empty lists called ask_posts, show_posts, and other_posts.
ask_post = []
show_post = []
other_post = []
# creating list of list for the different post titles
for row in hn:
    title = row[1]
    lower = title.lower()
    if lower.startswith("ask hn"):
        ask_post.append(row)
    elif lower.startswith("show hn"):
        show_post.append(row)
    else:
        other_post.append(row)
        
#printing the lenght of the posts
print(len(ask_post))
print(len(show_post))
print(len(other_post))
        

1744
1162
17194


The Ask HN post is 1,744 the Show HN is 1,162 in number and other post are 17,194. the Ask post are more than the show post.

## Calculating the Average Number Of Comments for Ask HN and Show HN Posts

In [6]:
# calculating the average comments for Ask HN
total_ask_comments = 0
for post in ask_post:
    comments = post[4]
    total_ask_comments += int(comments)
    
avg_ask_comments = total_ask_comments/ len(ask_post)
print(avg_ask_comments)

# calculating the average comments for Ask HN
total_show_comments = 0

for post in show_post:
    total_show_comments += int(post[4])
    
avg_show_comments = total_show_comments / len(show_post)
print(avg_show_comments)



14.038417431192661
10.31669535283993


The ask post on average has more comments than the show post.

## Finding the amount of ask posts and comments by hour created.

Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.
we'll determine if ask posts created at a certain time are more likely to attract comments. 

In [8]:
import datetime as dt

result_list = []
for post in ask_post:
    created_at = post[6]
    comments = post[4]
    result_list.append([created_at,int(comments)])
    
counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"
for row in result_list:
    date = row[0]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time not in counts_by_hour:
        counts_by_hour[time] = 1
        comments_by_hour[time] = row[1]
    else:
        counts_by_hour[time] += 1
        comments_by_hour[time] += row[1]
print(counts_by_hour)
print(comments_by_hour)

{'10': 59, '15': 116, '13': 85, '05': 46, '06': 44, '12': 73, '18': 109, '02': 58, '23': 68, '07': 34, '11': 58, '20': 80, '17': 100, '01': 60, '19': 110, '22': 71, '08': 48, '16': 108, '04': 47, '03': 54, '09': 45, '14': 107, '00': 55, '21': 109}
{'10': 793, '15': 4477, '13': 1253, '05': 464, '06': 397, '12': 687, '18': 1439, '02': 1381, '23': 543, '07': 267, '11': 641, '20': 1722, '17': 1146, '01': 683, '19': 1188, '22': 479, '08': 492, '16': 1814, '04': 337, '03': 421, '09': 251, '14': 1416, '00': 447, '21': 1745}


## Calculating the average number of comments for ask Hn post by hour

we'll use these two dictionaries to calculate the average number of comments for posts created during each hour of the day.

In [9]:
avg_by_hour = []
for hour in counts_by_hour:
    avg_by_hour.append([hour,comments_by_hour[hour]/counts_by_hour[hour]])
avg_by_hour

[['10', 13.440677966101696],
 ['15', 38.5948275862069],
 ['13', 14.741176470588234],
 ['05', 10.08695652173913],
 ['06', 9.022727272727273],
 ['12', 9.41095890410959],
 ['18', 13.20183486238532],
 ['02', 23.810344827586206],
 ['23', 7.985294117647059],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034],
 ['20', 21.525],
 ['17', 11.46],
 ['01', 11.383333333333333],
 ['19', 10.8],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['16', 16.796296296296298],
 ['04', 7.170212765957447],
 ['03', 7.796296296296297],
 ['09', 5.5777777777777775],
 ['14', 13.233644859813085],
 ['00', 8.127272727272727],
 ['21', 16.009174311926607]]

formating the output 

In [13]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])
    
print(swap_avg_by_hour)

#sort swap_avg_by_hour by average 
sorted_swap = sorted(swap_avg_by_hour,reverse = True)

print("Top 5 Hours for Ask Posts Comments")
for row in sorted_swap[:6]:
    hour = row[1]
    time = dt.datetime.strptime(hour, "%H").strftime("%H:%M:")
    print("{} {:.2f} average comments per post".format(time,row[0]))
    

[[13.440677966101696, '10'], [38.5948275862069, '15'], [14.741176470588234, '13'], [10.08695652173913, '05'], [9.022727272727273, '06'], [9.41095890410959, '12'], [13.20183486238532, '18'], [23.810344827586206, '02'], [7.985294117647059, '23'], [7.852941176470588, '07'], [11.051724137931034, '11'], [21.525, '20'], [11.46, '17'], [11.383333333333333, '01'], [10.8, '19'], [6.746478873239437, '22'], [10.25, '08'], [16.796296296296298, '16'], [7.170212765957447, '04'], [7.796296296296297, '03'], [5.5777777777777775, '09'], [13.233644859813085, '14'], [8.127272727272727, '00'], [16.009174311926607, '21']]
Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post
13:00: 14.74 average comments per post


from the above output we can see that the hours with higest averages are 15, 2, 20, 16 and 21. form the [documentation](https://www.kaggle.com/hacker-news/hacker-news-posts) the time zone used is Eastern Time in the US which is 5 hours behind Nigerian time. so in Nigerian time the best time to post ASK HN is by 8pm,7am,2am,9pm and 3am. 

## conclusion

In this project, we analyzed the Ask HN post and Show HN post of Hacker news, a platform where user-submitted stories (known as "posts") are voted and commented upon.Users submit Ask HN posts to ask the Hacker News community a specific question and Show HN posts to show the Hacker News community a project, product, or just generally something interesting.
     We found that on the average the Ask post has more comments than the Show post and when we went ahead to find the best time to post an Ask, we found that on a average the hour that has more comment was 8pm,7am,2am,9pm and 3am Nigerian time.